pandas/io/stata.py

```diff
@@ -951,222 +953,324 @@ def __init__(self, path_or_buf, convert_dates=True,
+        self.path_or_buf.read(1)  # unused
+
+        self.nvar = struct.unpack(self.byteorder + 'H',
+                                  self.path_or_buf.read(2))[0]
+        self.nobs = struct.unpack(self.byteorder + 'I',
+                                  self.path_or_buf.read(4))[0]
+        if self.format_version > 105:
+            self.data_label = self._null_terminate(self.path_or_buf.read(81))
+        else:
+            self.data_label = self._null_terminate(self.path_or_buf.read(32))
+        if self.format_version > 104:
+            self.time_stamp = self._null_terminate(self.path_or_buf.read(18))
 
-            if self.format_version > 104:
-                while True:
-                    data_type = struct.unpack(self.byteorder + 'b',
-                                              self.path_or_buf.read(1))[0]
-                    if self.format_version > 108:
-                        data_len = struct.unpack(self.byteorder + 'i',
-                                                 self.path_or_buf.read(4))[0]
-                    else:
-                        data_len = struct.unpack(self.byteorder + 'h',
-                                                 self.path_or_buf.read(2))[0]
-                    if data_type == 0:
-                        break
-                    self.path_or_buf.read(data_len)
+        # descriptors
+        if self.format_version > 108:
+            typlist = [ord(self.path_or_buf.read(1))
+                       for i in range(self.nvar)]
+        else:
+            typlist = [
+                self.OLD_TYPE_MAPPING[
+                    self._decode_bytes(self.path_or_buf.read(1))
+                ] for i in range(self.nvar)
+            ]
 
-            # necessary data to continue parsing
-            self.data_location = self.path_or_buf.tell()
+        try:
+            self.typlist = [self.TYPE_MAP[typ] for typ in typlist]
+        except:
+            raise ValueError("cannot convert stata types [{0}]"
+                             .format(','.join(typlist)))
+        try:
+            self.dtyplist = [self.DTYPE_MAP[typ] for typ in typlist]
+        except:
+            raise ValueError("cannot convert stata dtypes [{0}]"
+                             .format(','.join(typlist)))
 
-        self.has_string_data = len([x for x in self.typlist
-                                    if type(x) is int]) > 0
+        if self.format_version > 108:
+            self.varlist = [self._null_terminate(self.path_or_buf.read(33))
+                            for i in range(self.nvar)]
+        else:
+            self.varlist = [self._null_terminate(self.path_or_buf.read(9))
+                            for i in range(self.nvar)]
+        self.srtlist = struct.unpack(
+            self.byteorder + ('h' * (self.nvar + 1)),
+            self.path_or_buf.read(2 * (self.nvar + 1))
+        )[:-1]
+        if self.format_version > 113:
+            self.fmtlist = [self._null_terminate(self.path_or_buf.read(49))
```

Comment:
not sure if reading of some of the older formats can share some of these utilty functions

e.g. might be useful to make functions like these

```
def _get_format_list(self):
        if self.format_version > 113:
            return [self._null_terminate(self.path_or_buf.read(49))
                            for i in range(self.nvar)]
        elif self.format_version > 104:
            return [self._null_terminate(self.path_or_buf.read(12))
                            for i in range(self.nvar)]
        else:
            return [self._null_terminate(self.path_or_buf.read(7))
                            for i in range(self.nvar)]
```